### PR TITLE
PR - Iris Disbursement API

### PR DESCRIPTION
**Pull request by Xaxxis**
New feature:
- API Iris disbursement
- Payment BIN API
- Set Connection Pool & Keep alive connection
- Set header X-Idempotency-key for Iris and Idempotency-Key for Payment
- Set `X-Append-Notification`: to add new notification url(s) alongside the settings on dashboard
- Set `X-Override-Notification`: to use new notification url(s) disregarding the settings on dashboard
- Library version on `User-Agent` header

Improvement:

- Change logic for setup headers
- Remove versioning path from base URL, move to `String API_VERSION` in CoreAPI, SnapAPI interface class
- Improve conditional logic for HttpLoggingInterceptors. Now all debugging response handle only from HttpLoggingInterceptors Which can show logs request and response information

Bugs fix:
- Failure/error response from Midtrans api is empty json.

### CODE REVIEW

File: library/pom.xml [hunk 1]
```diff
@@ -80,13 +80,22 @@
...
+			</resource>
+		</resources>
 	</build>
 
 	<distributionManagement>
 		<repository>
 			<id>bintray-midtrans-midtrans-java</id>
 			<name>midtrans-midtrans-java</name>
-			<url>https://api.bintray.com/maven/midtrans/midtrans-java/com.midtrans/;publish=1</url>
+            <url>https://api.bintray.com/maven/midtrans/midtrans-java/com.midtrans/;publish=1</url>
```
Comment (rizdaprasetya) [inline library/pom.xml:98]: indentation looks not consistent

File: library/src/main/java/com/midtrans/Config.java [hunk 2]
```diff
@@ -8,19 +8,25 @@
 public class Config {
 
     private String SERVER_KEY;
-
     private String CLIENT_KEY;
-
     private boolean isProduction;
+    private String irisMerchantKey;
```
Comment (rizdaprasetya) [inline library/src/main/java/com/midtrans/Config.java]: Why is it different from `SERVER_KEY` to `irisMerchantKey`? not  uppercase?

File: library/src/main/java/com/midtrans/service/MidtransIrisApi.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+     */
+    JSONObject getFacilitatorBalance(String bankAccountId) throws MidtransError;
+
+    /**
+     * Do get `/beneficiary_banks` Use this API for show list of supported banks in IRIS. https://iris-docs.midtrans.com/#supported-banks
+     *
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response (more params detail refer to: https://iris-docs.midtrans.com/#list-banks)
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONObject getListBank() throws MidtransError;
```
Comment (rizdaprasetya) [inline library/src/main/java/com/midtrans/service/MidtransIrisApi.java]: change to more descriptive `getBeneficiaryBanks`

File: library/src/main/java/com/midtrans/service/MidtransIrisApi.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    JSONObject rejectPayouts(Map<String, Object> params) throws MidtransError;
+
+    /**
+     * Do get `/payouts/{reference_no}` Use this API for get details of a single payout
+     *
+     * @param referenceNo String parameter, unique reference no of a payout (more params detail refer to: https://iris-docs.midtrans.com/#get-payout-details)
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONObject getPayoutsDetails(String referenceNo) throws MidtransError;
```
Comment (rizdaprasetya) [inline library/src/main/java/com/midtrans/service/MidtransIrisApi.java]: `getPayoutDetails` payout without `s`

File: library/src/main/java/com/midtrans/service/MidtransIrisApi.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+     */
+    JSONObject updateBeneficiaries(String aliasName, Map<String, String> params) throws MidtransError;
+
+    /**
+     * Do get `/beneficiaries` Use this API to fetch list of all beneficiaries saved in Iris Portal.
+     *
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response refer to: https://iris-docs.midtrans.com/#list-beneficiaries
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONArray getListBeneficiaries() throws MidtransError;
```
Comment (rizdaprasetya) [inline library/src/main/java/com/midtrans/service/MidtransIrisApi.java]: change to `getBeneficiaries`

File: library/src/main/java/com/midtrans/service/MidtransIrisApi.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+     */
+    String ping() throws MidtransError;
+
+    /**
+     * Do get `/balance` API request to Use get current balance information. For Aggregator Partner you need to top up to Iris’ bank account.
+     *
+     * @return {JSONObject} - org.json Promise contains Object from JSON decoded response refer to: https://iris-docs.midtrans.com/#check-balance-aggregator
+     * @throws MidtransError when an exception was occurred during executing the request.
+     */
+    JSONObject getAggregatorBalance() throws MidtransError;
```
Comment (rizdaprasetya) [inline library/src/main/java/com/midtrans/service/MidtransIrisApi.java]: change to `getBalance`

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
This is technical feedback, readme related feedback ill be submitted separately

File: CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+- Set header X-Idempotency-key for Iris and Idempotency-Key for Payment
+- Set `X-Append-Notification`: to add new notification url(s) alongside the settings on dashboard
+- Set `X-Override-Notification`: to use new notification url(s) disregarding the settings on dashboard
+- Library version on `User-Agent` header
+
+Improvement:
+
+- Change logic for setup headers
+- Remove versioning path from base URL, move to `String API_VERSION` in CoreAPI, SnapAPI interface class
+- Improve conditional logic for HttpLoggingInterceptors. Now all debugging response handle only from HttpLoggingInterceptors Which can show logs request and response information
```
Comment (rizdaprasetya) [inline CHANGELOG.md]: `will be handled only from`

File: CHANGELOG.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+Bugs fix:
+- Failure/error response from Midtrans api is empty json.
```
Comment (rizdaprasetya) [inline CHANGELOG.md]: `Midtrans api result in empty json`

File: CHANGELOG.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+- Improve conditional logic for HttpLoggingInterceptors. Now all debugging response handle only from HttpLoggingInterceptors Which can show logs request and response information
+
+Bugs fix:
+- Failure/error response from Midtrans api is empty json. 
+
+## v1.1.1 (Feb 3, 2020)
+
+Improvement:
+
+HttpLoggingInterceptor: Which can show logs request and response information. This feature auto disable when using production mode.
```
Comment (rizdaprasetya) [inline CHANGELOG.md]: This feature is disabled by default on production mode.

File: CHANGELOG.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+## v1.1.0 (Nov 27, 2019)
+
+Feature:
+
+- Can use HTTP Proxy feature
+- Can set connectionTimeout
+- Can set readTimeout
+- Can set writeTimeout
```
Comment (rizdaprasetya) [inline CHANGELOG.md]: - Allow HTTP Proxy config
- Allow set connectionTimeout
- Allow set readTimeout
- Allow set writeTimeout

File: README.md [hunk 11]
```diff
@@ -87,7 +88,20 @@ MidtransSnapApi snapApi = new ConfigFactory(new Config("YOU_SERVER_KEY","YOUR_CL
...
+
+MidtransIrisApi irisApi = new ConfigFactory(new Config("IRIS-CREDENTIALS",null , false)).getIrisApi();
+
+//You can set Config.class with
+`IRIS-CREDNTIALS`
+`null`
+`isProduction`
+```
+
+You can also re-setting config using `apiConfig()` method on MidtransCoreApi.Class, MidtransSnapApi.Class or MidtransIrisApi.class like `coreApi.apiConfig.set( ... )`
```
Comment (rizdaprasetya) [inline README.md]: can also re-set config

File: README.md [hunk 12]
```diff
@@ -99,9 +113,15 @@ coreApi.apiConfig().setSERVER_KEY("YOUR_SERVER_KEY");
...
+
+// For Iris Disbursement can set creator & approver credentials with apiConfig()
+irisApi.apiConfig().setSERVER_KEY("IRIS-CREDENTIALS");
+
+irisApi.apiConfig().setIrisIdempotencyKey("IRIS-IDEMPOTENCY-KEY");
+
 ```
 
-In production environment, LOG is automatically turn off, if you want enable LOG, you can turn on log with this
+In production environment, LOG is automatically turn off, if you want enable LOG, you can `setEnabledLog` like this
```
Comment (rizdaprasetya) [inline README.md]: `LOG is by default turned off`
`you can enable by setEnabledLog, e.g:`

File: README.md [hunk 13]
```diff
@@ -514,12 +557,192 @@ private Map<String,String> params() {
 JSONObject refundTransaction("YOUR_ORDER_ID OR TRANSACTION_ID", params());
     //do something with `result` JSON Raw Object
 ```
+### 2.2.D Iris API (Disbursement)
+You can see some Iris API examples [here](example/src/main/java/com/midtrans/sample/controller/IrisController.java).
```
Comment (rizdaprasetya) [inline README.md]: remove `some`

File: library/src/main/java/com/midtrans/service/MidtransIrisApi.java [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+import com.midtrans.httpclient.error.MidtransError;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.Map;
+
+public interface MidtransIrisApi {
+
+    /**
+     * Do re-setting config Class iris-credential, IrisIdempotencyKey
```
Comment (rizdaprasetya) [inline library/src/main/java/com/midtrans/service/MidtransIrisApi.java]: `re-set config Class`

**Review by rizdaprasetya — CHANGES_REQUESTED** [review]
Feedback for readme related

File: library/pom.xml (hunk 1 above)
Comment (Xaxxis) [reply library/pom.xml:98]: done

**Review by rizdaprasetya — APPROVED** [review]
Looks good now! please merge and release! Thanks @Xaxxis

**Pull request merged**